### ISSUE TITLE
 NegativeArraySizeException on saving

### ISSUE THREAD

**Issue by KLemke-85** [JabRef/jabref#8265]
### JabRef version

5.3 (latest release)

### Operating system

Windows

### Checked with the latest development build

- [X] I made a backup of my libraries before testing the latest development version.
- [X] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

1. System fails to save updated entries (annotation and correcting commata etc.). Closing the program after that a big amount of data (400 of over 900 entries) was deleted, even though the user had not worked on those entires, and the subcategories vanished completely. The same issue occured before and after the updated version. Data could be restored due to externally made backups.
2. User is saving entries frequently. But even though older entries got deleted.
3. No particular occurances or steps before the error is reported.
4. The attempt to export the existing data after that occurance failed as well. Only a small amount was exported for example in rtf. Each try provided different data sets, but never all existing entries.

### Appendix

...
<details>

<summary>Log File</summary>

```
2 screenshots of the error notification.
```

</details>
<img width="413" alt="JabRef 2021_11_18" src="https://user-images.githubusercontent.com/72263522/142839344-759ee01c-814e-402b-bbe8-ef9e0c06354d.PNG">
![JabRef 2021_11_19](https://user-images.githubusercontent.com/72263522/142839360-885288ee-3242-4212-b91c-3b5396230fe5.JPG)

**Comment by ThiloteE** [comment]
Uff, this sounds bad, but as a fellow user (not programmer) of Jabref who is trying to help in debugging this, i do have some follow up questions, to make everything a little bit more clear:

-  "fails to save updated entries (annotation and correcting commata etc.)"?
In particular, were there or were there not any changes to annotations and commas in the entries (manually or automated?) prior to closing Jabref? The above sentence seems to contradict your later statement of no particular steps being conducted before the error, so clarification would be VERY helpful for debugging. What exactly did you do to annotations and or commas?
- With "updated version" you mean the latest development version? Could you please post the detailed Jabref version you are using? You can find it under `Help>About Jabref`. This also would answer if this is Windows 10 or Windows 7.
- When you open your library-file with a text editor (e.g. `right-click>open with>editor`) can you still see the entries that have vanished from Jabref? Are the entries gone from the file completely or are they just not depicted within Jabref?
- Is the data of sensitive nature? If not, could you upload both the file PRIOR to saving and the file(s) that were created AFTER the saving here on github for the community to be able to debug it?
- Are you using manual or the autosave function? (find it under: `options>preferences>file>autosave`. If you use autosave, try to deactivate it please.

In general, next time you encounter this or any problems, a full copy of the debuglog would be more adviced than a screenshot. you can capture it on windows machines by clicking into the error message, then press ctrl+A (which selects all text) and then press ctrl+c (copy) and then you can paste it on github or wherever you see fit.

**Comment by ThiloteE** [comment]
Thanks for the immediate response as well.
- ok, so what then was the older Jabref version you upgraded from? :D

What you report is very serious, but i am very pessimistic this can be solved without knowing what exactly changed in the files and/or any steps that allow us to reproduce this issue. I fear there is just not enough data to investigate this, but If you updated from a very old version of Jabref to the newer 5.3 version, maybe what was encountered is a known problem. For that we would need to know the exact version of your old Jabref.

The only other thing i can think of is for the experts to stare at the jabref code real long and hard and start to think how such an error could come into being, but that is like trying find a needle in a haystack....

**Comment by Siedlerchr** [comment]
Thank you very much for your report.  The exception looks similar to #8231  A quick look in the code seems to show that it has to do with writing the formatted field names. 
I guess that somehow the field.length is wrong/or negative (probaby). 
Indentation (the second parameter) is the length of the longest field name in the library - or zero)

https://github.com/JabRef/jabref/blob/5680d154d238e4dc3d9cc76968a490e30dca4798/src/main/java/org/jabref/logic/bibtex/BibEntryWriter.java#L191-L193 
Do you use any Custom entry types or entry types with (non standard) fields? 

PS: You can send bib files with sensitive data to vorstand[at]jabref.org or developer[at]jabref.org. They will be handled securely by the dev team only.

**Comment by KLemke-85** [comment]
Thank you for pointing that out. Well, I'm currently not aware of such a customization. Therefore I only can assume that it has to do with the field length or indentation respectively. We obviously must have overlooked something.
I will send you the file via the suggested address.

**Comment by KLemke-85** [comment]
And to answer your question @ThiloteE 
I don't recall the version number, bit we only indroduced Jabref in the end of 2020, latest in the beginning of 2021. It cannot be that old then.

**Comment by koppor** [comment]
I should finish https://github.com/JabRef/blog.jabref.org/pull/47 ^^

**Comment by koppor** [comment]
@KLemke-85 Could you please try the binary at @AlbertJBurton Could you please try the binary at https://builds.jabref.org/pull/8270/merge/? In case you see some wrong indent at some bib entry, it would be nice if you cold post that bibentry here for further investigation.

**Comment by KLemke-85** [comment]
I'm very sorry, but I have no idea, what I'm supposed to do with that, literally. And I was trying to send our bib-file to your developer address, but the mail is bouncing back.

**Comment by Siedlerchr** [comment]
@KLemke-85  There was a typo: try developers@jabref.org (there was an "s" missing)

**Comment by KLemke-85** [comment]
okeee... now I can see the s... Thx!

**Comment by Siedlerchr** [comment]
Thanks for the file. If you open the file with a text editor ( Editor on Windows or any other plain text editor)  and scroll down to the end, after the comment section, there are some errors. Delete that "Schlemmer1927" and the line below.

And also delete the character "**q**" from the line:  `@Comment{jabref-meta: grouping: q`
Afterwards, the file should be working again. At least I could open it.

However, I noticed some weird things regarding the authors. Some authors have square Brackets in the name...

**Comment by KLemke-85** [comment]
Ok. I will look into that :) thank you very much!!!
And the square brackets actually have a purpose; my colleague is using them to differentiate between given author names and names, were originally only initials are given and we have to provide the actual name. Do you think, that this also could cause complications? I was wondering about it for quite some time. But as mentioned above, we didn't have any issues until recently.
But maybe its the "Schlemmer1927" after all, a complaint, that I didn't know how to come by yet. :)

**Comment by KLemke-85** [comment]
I deleted the Schlemmer-Error (finally), but didn't have a "q" in the mentioned line. Should be a bother. I guess.
I assume, we will have to wait now and see how it turns out, when my colleague is working on the Bib again. But now I know where to look for a start, if there are any hickups.
Thank you very much!

**Comment by koppor** [comment]
With JabRef portable of v5.3, I get

![grafik](https://user-images.githubusercontent.com/1366654/143133610-d05f291e-02f5-48d3-b6a8-19fcdd34f608.png)

Saving works. Strange.

**Comment by KLemke-85** [comment]
Yes, that is correct. I already have dealt with that, thx to @Siedlerchr and managed to delete that warning/entry. We had this warning for a while, in the beginning with a different entry. But I didn't know how to treat it at the time. But there never was a problem with saving and such. Therefore we ignored it... we probably shouldn't have.

And from the little experience I have, while using the portable version an error like the one I opened the issue for didn't come up (yet).

**Comment by KLemke-85** [comment]
Unfortunately there was a new warning, but now I can provide the complete bug report, which you may find below. Just from this morning. I also attached to screenshots, one shows a new warning.


java.lang.NegativeArraySizeException: -2
	at java.base/java.lang.AbstractStringBuilder.<init>(Unknown Source)
	at java.base/java.lang.StringBuilder.<init>(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.model.strings.StringUtil.repeat(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.model.strings.StringUtil.repeatSpaces(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.logic.bibtex.BibEntryWriter.getFormattedFieldName(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.logic.bibtex.BibEntryWriter.writeField(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.logic.bibtex.BibEntryWriter.writeRequiredFieldsFirstRemainingFieldsSecond(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.logic.bibtex.BibEntryWriter.write(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.logic.exporter.BibtexDatabaseWriter.writeEntry(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.logic.exporter.BibDatabaseWriter.savePartOfDatabase(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.logic.exporter.BibDatabaseWriter.saveDatabase(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.gui.exporter.SaveDatabaseAction.saveDatabase(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.gui.exporter.SaveDatabaseAction.save(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.gui.exporter.SaveDatabaseAction.save(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.gui.exporter.SaveDatabaseAction.save(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.gui.exporter.SaveAction.execute(Unknown Source)
	at org.jabref@5.3.60000/org.jabref.gui.actions.JabRefAction.lambda$new$3(Unknown Source)
	at org.jabref.merged.module@5.3.60000/org.controlsfx.control.action.Action.handle(Unknown Source)
	at org.jabref.merged.module@5.3.60000/org.controlsfx.control.action.Action.handle(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventUtil.fireEventImpl(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventUtil.fireEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/javafx.event.Event.fireEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/javafx.scene.Node.fireEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/javafx.scene.control.Button.fire(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.scene.control.inputmap.InputMap.handle(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventUtil.fireEventImpl(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.event.EventUtil.fireEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/javafx.event.Event.fireEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/javafx.scene.Scene$MouseHandler.process(Unknown Source)
	at org.jabref.merged.module@5.3.60000/javafx.scene.Scene.processMouseEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/javafx.scene.Scene$ScenePeerListener.mouseEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$2(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.glass.ui.View.handleMouseEvent(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.glass.ui.View.notifyMouse(Unknown Source)
	at org.jabref.merged.module@5.3.60000/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at org.jabref.merged.module@5.3.60000/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)

![JabRef 2021_11_24](https://user-images.githubusercontent.com/72263522/143221071-5745b1dd-6fb9-4a28-a6ec-b18382c1d171.JPG)

![JabRef 2021_11_24 a](https://user-images.githubusercontent.com/72263522/143221092-d63ba0d8-cfdc-49c2-90e2-70367c9f0656.JPG)

**Comment by ThiloteE** [comment]
I think 'Error in line 7007' means that the problem is in row 7007 of your library-file. No?

**Comment by Siedlerchr** [comment]
Yes, that is the line where it failed, the root cause of the error could be in lines earlier. It has something to do with the braces. This is often the case where a closing curly brace "}" is missing and so JabRef thinks the entry's field continues in the following lines

**Comment by KLemke-85** [comment]
I just checked the number of braces, all accounted for it seems :) but I've  noticed, that the output of entries with the additional information of the month of publication are not representated with a curly brace, if I open the data via text editor. And I found a couple of square brackets that are not individually cupped by curly braces.

And, even though I've applied what you've sugggested yesterday @Siedlerchr the warning "schlemmer1927" turned up again today.
And I have to correct myself. Though we already had installed the portable version as well, I was not using it. The different warnings on my colleague's computer are still puzzeling. Even more so.
I also remember that I checked the use of square brackets in title entries, if using and additional set of curly braces, a couple of months ago. Still, I don't know if the warning really is related to that.
